OpenRA.Mods.D2k/Traits/Thumper.cs

```diff
@@ -0,0 +1,98 @@
+#region Copyright & License Information
+/*
+ * Copyright 2007-2015 The OpenRA Developers (see AUTHORS)
+ * This file is part of OpenRA, which is free software. It is made
+ * available to you under the terms of the GNU General Public License
+ * as published by the Free Software Foundation. For more information,
+ * see COPYING.
+ */
+#endregion
+
+using System.Collections.Generic;
+using System.Linq;
+using OpenRA.Mods.Common.Orders;
+using OpenRA.Mods.Common.Traits;
+using OpenRA.Traits;
+
+namespace OpenRA.Mods.D2k.Traits
+{
+	[Desc("Lures the `Sandworm` to its position when deployed, but doesn't attract it to avoid becoming a target.")]
+	class ThumperInfo : ITraitInfo, Requires<MobileInfo>
+	{
+		[SequenceReference(null, true)] public readonly string UndeployedPrefix = "undeployed-";
+
+		public readonly int ThumpInterval = 40;
+		public readonly string ThumpSound = "THUMPER1.WAV";
+
+		public readonly string[] DeployTerrainTypes = { "Sand", "Dune", "Spice" };
+		public readonly string DeployCursor = "deploy";
+		public readonly string DeployBlockedCursor = "deploy-blocked";
+
+		[VoiceReference] public readonly string Voice = "Action";
+
+		public object Create(ActorInitializer init) { return new Thumper(init.Self, this); }
+	}
+
+	class Thumper : IRenderInfantrySequenceModifier, ITick, IIssueOrder, IResolveOrder, IOrderVoice
```

Comment:
You can then avoid almost all of the duplication in this trait by making it subclass `DeployToUpgrade` and add the `IRenderInfantrySequenceModifier` and sound code on top.  The `AttractsWorms` trait already supports using upgrades to switch to a version with much higher attraction when deployed.